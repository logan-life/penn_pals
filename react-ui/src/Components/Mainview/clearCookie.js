export default async function clearCookie() {
  try {
    const response = await fetch("api/logout", {
      method: "GET",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });

    if (response.status === 200) {
      //removing username from localStorage
      localStorage.removeItem("currentUser");
      return response.status;
    } else {
      const error = new Error(response.error);
      throw error;
    }
  } catch (err) {
    console.error(err);
    // document.getElementById("statusMain").innerHTML = err;
  }
}
