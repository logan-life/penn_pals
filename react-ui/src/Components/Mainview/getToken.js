/**
 *
 * @param {string} username - this is the linkage between PennPals and Twilio. username for Pennpals === identity in Twilio
 */

export default async function getToken(username) {
  try {
    const token = await fetch(`api/twilio?username=${username}`, {
      method: "GET",
      credentials: "include",
      headers: {
        "Content-Type": "application/json",
      },
    });
    return token.text();
  } catch (error) {
    throw new Error(error);
  }
}
