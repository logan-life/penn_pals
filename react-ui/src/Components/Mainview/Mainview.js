import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";
import Spinner from "react-bootstrap/Spinner";
import { useHistory } from "react-router-dom";

import { Link } from "react-router-dom";
// import "./Contacts/ContactView.css";
import { initializeConversationsClient } from "../Mainview/Contacts/contactHelperFunctions.js";
import "./assets/mainview.css";
import React, { useEffect, useState } from "react";

import ContactView from "./Contacts/ContactView";
import clearCookie from "./clearCookie";
import getToken from "./getToken";

export default function Mainview() {
  const history = useHistory();
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [username, setUsername] = useState(null);
  const [twilioToken, setTwilioToken] = useState(null);

  const [client, setClient] = useState(null);

  const [clientConnectionState, setClientConnectionState] = useState(null);
  const [clientReady, setClientReady] = useState(false);

  //checks localStorage to see if user is logged in, updates state accordingly.
  useEffect(() => {
    let currentUser = localStorage.getItem("currentUser");
    if (currentUser) {
      setUsername(currentUser);
      setIsLoggedIn(true);
    } else {
      setUsername(null);
      setIsLoggedIn(false);
    }
  }, []);

  //gets a token for a logged in user and instantiates the twilio Client library instance
  useEffect(() => {
    if (isLoggedIn && !client) {
      getToken(username).then((token) => {
        setTwilioToken(token);
        if (token) {
          initializeConversationsClient(token).then((res) => {
            setClient(res);
            if (client) {
              setClientConnectionState(res.connectionState);
            }
          });
        }
      });
    }
  });

  //subscribes to the client connection state, listening for client to finish connecting
  useEffect(() => {
    if (client) {
      client.on("connectionStateChanged", (state) => {
        setClientConnectionState(state);
      });
      if (clientConnectionState === "connected") {
        setClientReady(true);
      }
    }
  }, [client, clientConnectionState]);

  //shuts down the client and resets state
  function handleLogOut() {
    if (client) {
      client.shutdown().then((res) => {
        setClient(null);
        setClientConnectionState(null);
        setClientReady(false);
      });
    }
    clearCookie();
    setIsLoggedIn(false);
    setTwilioToken(null);
    setUsername(null);
    history.push("/login");
  }

  const mainViewDiv = (
    <div className="mainviewTitleAndLogoutDiv">
      <h1 id="statusMain">Welcome!</h1>

      <div className="Logout">
        <Button id="logoutBtn" variant="primary" onClick={handleLogOut}>
          Logout
        </Button>
      </div>
    </div>
  );

  return !isLoggedIn ? (
    <Alert id="alertContactViewNotLoggedIn" variant="danger">
      Please{" "}
      <Alert.Link as={Link} to="/Login">
        login
      </Alert.Link>{" "}
      to see your contacts.
    </Alert>
  ) : isLoggedIn && !clientReady ? (
    <Spinner animation="border" role="status" variant="primary">
      <span className="sr-only">Loading...</span>
    </Spinner>
  ) : (
    <div>
      <div className="mainviewWrapperDiv">
        {mainViewDiv}
        <ContactView
          username={username}
          twilioToken={twilioToken}
          client={client}
        />
      </div>
    </div>
  );
}
