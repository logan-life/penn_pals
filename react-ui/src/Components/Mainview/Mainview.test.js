import React from 'react';
import Mainview from './Mainview';
import {BrowserRouter} from "react-router-dom";
const renderer = require('react-test-renderer');
let container;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  document.body.removeChild(container);
  container = null;
});

// snapshot testing
test('Mainview view matches snapshot', () => {
const component = renderer.create(<BrowserRouter>
          <Mainview />
        </BrowserRouter>
      );
let tree = component.toJSON();
expect(tree).toMatchSnapshot();
});