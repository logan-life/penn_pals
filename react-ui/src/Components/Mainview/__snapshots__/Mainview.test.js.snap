// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mainview view matches snapshot 1`] = `
<div
  className="fade alert alert-danger show"
  id="alertContactViewNotLoggedIn"
  role="alert"
>
  Please
   
  <a
    className="alert-link"
    href="/Login"
    onClick={[Function]}
  >
    login
  </a>
   
  to see your contacts.
</div>
`;
